internal/worker/secretspruner/worker.go

```diff
@@ -78,16 +83,54 @@ func (w *Worker) loop() (err error) {
 		select {
 		case <-w.catacomb.Dying():
 			return errors.Trace(w.catacomb.ErrDying())
-		// TODO: watch for secret's auto-prune config changes.
-		// then delete any obsolete revisions.
-		case _, ok := <-watcher.Changes():
+		case changes, ok := <-watcher.Changes():
 			if !ok {
 				return errors.New("secret prune changed watch closed")
 			}
-			w.config.Logger.Debugf("maybe have user secret revisions to prune")
-			if err := w.config.SecretsFacade.DeleteObsoleteUserSecrets(); err != nil {
+			w.config.Logger.Debugf("got user supplied secret revisions to prune")
+
+			if len(changes) == 0 {
+				w.config.Logger.Debugf("no secret revisions to prune")
+				continue
+			}
+			revisions, err := w.processChanges(changes...)
+			if err != nil {
 				return errors.Trace(err)
 			}
+			for uriStr, revs := range revisions {
+				w.config.Logger.Debugf("pruning secret revisions %q: %v", uriStr, revs.SortedValues())
+				uri, err := secrets.ParseURI(uriStr)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if err := w.config.SecretsFacade.DeleteObsoleteUserSecrets(uri, revs.SortedValues()...); err != nil {
+					return errors.Trace(err)
+				}
+			}
+		}
+	}
+}
+
+func (w *Worker) processChanges(changes ...string) (map[string]set.Ints, error) {
+	out := make(map[string]set.Ints)
+	for _, revInfo := range changes {
+		w.config.Logger.Warningf("revInfo %q, out %#v", revInfo, out)
+		parts := strings.Split(revInfo, "/")
+		uri := parts[0]
+		if len(parts) < 2 {
```

Comment:
This isn't intuitive and seems wrong. Why would the watcher emit a revision URI for a deleted secret.